Addition of P.V. Refit after Tracking in Heavy Ion Reco

abaty opened the pull request:
This PR is aimed at drastically improving the PV resolution in the heavy ion reconstruction by adding an additional fitting step after the tracking.  Previously, the PV was formed using a subset of pixel tracks formed before the tracking is run, in the `hiSelectedVertex` collection.  

This PR does the following:
1) Changes the name of the `hiSelectedVertex` collection to `hiSelectedPixelVertex`
2) Changes all heavy ion tracking to pick up `hiSelectedPixelVertex` while running the tracking
3) After tracking, runs a copy of the pixel track vertexing after the tracking, but using `hiGeneralTracks` as the input collection
4) Filters the output to make sure 1 best vertex is produced, and stores that as the new `hiSelectedVertex` which is used by downstream products

The effects on physics performance in 2015 data have been studied  [here](https://indico.cern.ch/event/638128/contributions/2585615/attachments/1460150/2255038/primaryVertexRefit.pdf) with a similar setup.  In general, we see drastically improved vertex resolutions. This PR applies an extra nHit requirement on tracks in order to reduce timing requirements.  

The memory of the new vertexing step seems reasonable in our tests.  We have added `hiSelectedPixelVertex` (which contains only 1 vertex) to the event content, so a very small increase in the output storage size is required.  The timing seems reasonable, on the order of a 5% increase for central events.  This change should affect both Phase-0 and Phase-1 reconstructions.  Tested with WF 140.53, as well as the instructions [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/HiRecoTaskForce2018).

@mandrenguyen

Commit:
added additional HI vertexing step after tracking

Commit:
changed hiOfflinePrimaryVertex to distinguish between phase0,phase1 when cutting on nPix

cmsbuild commented:
A new Pull Request was created by  @abaty  for master.

It involves the following packages:

RecoHI/HiJetAlgos
RecoHI/HiMuonAlgos
RecoHI/HiTracking


@perrotta, @cmsbuild, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@yslai, @miheejo, @jazzitup, @echapon, @yenjie, @kurtejung, @mandrenguyen, @dgulhan, @yetkinyilmaz this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8c9ea2123948b7136099190b7674b4086faa0587
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19083/20315/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19083/20315/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3439 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1837639
 - DQMHistoTests: Total failures: 19821
 - DQMHistoTests: Total nulls: 75
 - DQMHistoTests: Total successes: 1817570
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

slava77 commented on RecoHI/HiTracking/python/HISelectedVertex_cfi.py:
```diff
@@ -7,11 +7,57 @@
 )
```
this file should be renamed to a cff.py
the name can be bestHiVertexSequences_cff.py or something similar
"HISelectedVertex" does not represent the content anymore.

abaty replied:
File was renamed to HIBestVertexSequences_cff.py

slava77 commented on RecoHI/HiTracking/plugins/HIBestVertexProducer.cc:
```diff
@@ -19,6 +19,8 @@ HIBestVertexProducer::HIBestVertexProducer
   theMedianVertexCollection(consumes<reco::VertexCollection>(ps.getParameter<edm::InputTag>("medianVertexCollection"))),
   theAdaptiveVertexCollection(consumes<reco::VertexCollection>(ps.getParameter<edm::InputTag>("adaptiveVertexCollection")))
 {
+  theUseFinalAdapativeVertexCollection  = (ps.existsAs<bool>("useFinalAdapativeVertexCollection") ? ps.getParameter<bool>("useFinalAdapativeVertexCollection") : false);
```
existsAs should normally not be used.
Please propagate this parameter to relevant configs.
Better yet, please define fillDescriptions method and define the value of this and other parameters there.
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp

abaty replied:
I have added a fillDescriptions function.

slava77 commented on RecoHI/HiTracking/plugins/HIBestVertexProducer.cc:
```diff
@@ -53,68 +55,93 @@ void HIBestVertexProducer::produce
+    ev.getByToken(theFinalAdaptiveVertexCollection, vc0);
+    const reco::VertexCollection *vertices0 = vc0.product();
+    if(vertices0->size()==0)
+      LogError("HeavyIonVertexing") << "final adaptive vertex collection is empty!" << endl;
```
How abnormal is this?
If minbias data has this every other event, this message should be changed to LogInfo.
If this is not really ever supposed to happen, it may be a configuration problem and it should be changed to an Exception.

abaty replied:
I think this is expected to happen occasionally in peripheral events which have only a few tracks, if some of them don't make it past the number of hits cut or into the tracker acceptance.  I have demoted from LogError to LogInfo.

slava77 commented on RecoHI/HiTracking/plugins/HIBestVertexProducer.cc:
```diff
@@ -53,68 +55,93 @@ void HIBestVertexProducer::produce
+  //if using final vertex and has a good vertex, use it
+    if(vertices0->begin()->zError()<3){
+      hasFinalVertex = true;
+      reco::VertexCollection::const_iterator vertex0 = vertices0->begin();
```
```auto cont& vertex0 vertices0->front();``` appears to be more appropriate (the iterator functionality is not needed here).

abaty replied:
Implemented.

slava77 commented on RecoHI/HiTracking/plugins/HIBestVertexProducer.cc:
```diff
@@ -53,68 +55,93 @@ void HIBestVertexProducer::produce
+      LogError("HeavyIonVertexing") << "final adaptive vertex collection is empty!" << endl;
+
+  //if using final vertex and has a good vertex, use it
+    if(vertices0->begin()->zError()<3){
```
should this magic number of 3 be made configurable?
Since now it is needed in two places in this file, please define a named constant and use it in both places.

abaty replied:
Implemented.

slava77 commented on RecoHI/HiTracking/python/HISelectedVertex_cfi.py:
```diff
@@ -7,11 +7,57 @@
+        maxNormalizedChi2 = cms.double(5.0),
+        minPixelLayersWithHits=cms.int32(3),    #0 missing pix hit
+        minSiliconLayersWithHits = cms.int32(5),#at least 8 hits total
+        maxD0Significance = cms.double(3.0),    #default 5.0, suppresses split vtxs
```
The text in the comments appears to be out of sync from the values.
Please update or clarify.

abaty replied:
I have added some clarification.

slava77 commented on RecoHI/HiTracking/python/HISelectedVertex_cfi.py:
```diff
@@ -7,11 +7,57 @@
+)
+from Configuration.Eras.Modifier_trackingPhase1_cff import trackingPhase1
+trackingPhase1.toModify(hiOfflinePrimaryVertices,
+TkFilterParameters = cms.PSet(
```
shouldn't this be just 
```python
trackingPhase1.toModify(hiOfflinePrimaryVertices, 
    TkFilterParameters = dict (minPixelLayersWithHits = 4, minSiliconLayersWithHits = 4, )
)
```

abaty replied:
Implemented.

slava77 commented on RecoHI/HiTracking/python/HISelectedVertex_cfi.py:
```diff
@@ -7,11 +7,57 @@
+    adaptiveVertexCollection = cms.InputTag("hiBestAdaptiveVertex"),
+    useFinalAdapativeVertexCollection = cms.bool(True),
+    finalAdaptiveVertexCollection = cms.InputTag("hiBestOfflinePrimaryVertex")
+)
```
to reduce copy paste, it seems better to have
```python
hiSelectedVertex = hiSelectedPixelVertex.clone(
      useFinalAdapativeVertexCollection = cms.bool(True),
      finalAdaptiveVertexCollection = cms.InputTag("hiBestOfflinePrimaryVertex")
)
```

abaty replied:
Implemented.

slava77 commented on RecoHI/HiTracking/python/HISelectedVertex_cfi.py:
```diff
@@ -7,11 +7,57 @@
+    useFinalAdapativeVertexCollection = cms.bool(True),
+    finalAdaptiveVertexCollection = cms.InputTag("hiBestOfflinePrimaryVertex")
+)
+bestFinalHiVertex = cms.Sequence(hiOfflinePrimaryVertices * hiBestOfflinePrimaryVertex * hiSelectedVertex ) # vertexing run BEFORE tracking
```
looks like the comment should say "AFTER tracking"

abaty replied:
Fixed.

slava77 commented on RecoHI/HiTracking/python/RecoHiTracker_EventContent_cff.py:
```diff
@@ -47,6 +49,7 @@
                                            'keep recoTracks_hiGeneralAndPixelTracks_*_*',
-                                           'keep recoVertexs_hiSelectedVertex_*_*'		
+                                           'keep recoVertexs_hiSelectedVertex_*_*',		
+                                           'keep recoVertexs_hiSelectedPixelVertex_*_*'		
```
at least for AOD: what is the purpose to save the pre-tracking vertex (hiSelectedPixelVertex)?
Will it be of any use in analyses?

abaty replied:
This is a good point, comparisons between the two collections can probably be done with a small sample of RECO.  I have removed it from the AOD, but left it in RECO.

slava77 commented:
@abaty 
sorry for the delay in the review.
I am now done with visual inspection.
Please follow up on the comments.
Thank you.

slava77 commented:
Looking at comparisons with this PR (in red) compared to CMSSW_9_2_2 as a baseline, there are roughly expected changes:

The vertex is more precise
<img width="837" alt="wf140 53_hivtx_xerr" src="https://user-images.githubusercontent.com/4676718/27149725-6331cee6-50f9-11e7-8153-56b81a96a2c6.png">

vertexChi2 plots have many entries going off-scale
<img width="837" alt="wf140 53_hivtx_chi2" src="https://user-images.githubusercontent.com/4676718/27149765-8467bdaa-50f9-11e7-947f-9d90b4ed006b.png">

Strangely, the vertex chi2 prob has a peak at 1
<img width="841" alt="wf140 53_hivtx_chi2prob" src="https://user-images.githubusercontent.com/4676718/27149795-9d1240be-50f9-11e7-872e-8ff6588d96eb.png">
perhaps there are some numerical issues in setting the vertex chi2:  chi2/ndof is typically around 0.90-0.95 for ndof of O(1000).

One striking difference in physics objects is that there is a large reduction in the number of electrons in the particleFlow collection. I'm guessing that they are now killed by isolation or track multiplicity requirements. Recall that the vertex size before this change was 30-50, compared to 3K tracks now. This apparently throws off the selections in PFAlgo
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/RecoParticleFlow/PFProducer/src/PFEGammaFilters.cc#L130
This may need to be fine-tuned.
OTOH, if someone is relying on electrons in HI reco, is it safe to assume they are using the GED collections (instead of picking electrons from particleFlowTmp collection)?

Commit:
change of name and cleanup of HISelectedVertex .py

Commit:
removed SelectedPixelVertex from AOD content (still in RECO)

Commit:
changes to HIBestVertexProducer and adding fillDescription

cmsbuild commented:
Pull request #19083 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

abaty commented:
@slava77 I have finished addressing the code comments from your inspection a few days ago.

I am a bit puzzled by the chi2 peak at 1, from my previous checks the chi2/ndof did seem to peak around 0.9.  Which workflow did you run, so that I can try to reproduce this?

@kurtejung Can you comment on if the issue described above for electrons in PF could be an issue for the PF jets?

slava77 commented:
On 6/15/17 1:26 PM, abaty wrote:
> @slava77 <https://github.com/slava77> I have finished addressing the
> code comments from your inspection a few days ago.
> 
> I am a bit puzzled by the chi2 peak at 1, from my previous checks the
> chi2/ndof did seem to peak around 0.9. Which workflow did you run, so
> that I can try to reproduce this?

I ran 140.53

If you confirm that you saw chi2/ndof peak around 0.9, then it's
expected to have chi2Prob peak at 1:
for nDof=3000 normal chi2/ndof should have a mean of 1.0 and a width of
0.02.
The value of chi2/ndof=0.9 is then 5 sigma below the normal mean and
prob for it is ~1.

> 
> @kurtejung <https://github.com/kurtejung> Can you comment on if the
> issue described above for electrons in PF could be an issue for the PF jets?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/19083#issuecomment-308856402>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbvvkfsMr3999zPP7wJwy7zGcY7uqks5sEZORgaJpZM4NugHs>.
>

kurtejung commented:
@abaty - I think the electron reco is a question for @ttrk... AFAIK, we still use the GED photons for analysis, but the particleFlowTmp collections are used for jet clustering.  e's and mu's are fairly rare in jets (~10^-3 less common than calo hits & trks) so it should not dramatically affect the performance but would be good to understand the reason for the efficiency drop

ttrk commented:
@slava77 Electrons used in HI analysis are GED collections. But, I guess a problem in PF collection would affect the GED as well, because PF candidates are input for the latter. Is that right ?

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
On 6/15/17 7:30 PM, ttrk wrote:
> @slava77 <https://github.com/slava77> Electrons used in HI analysis are
> GED collections. But, I guess a problem in PF collection would affect
> the GED as well, because PF candidates are input for the latter. Is that
> right ?

The "early" GED electrons ("tmp" in pp sequence) are made before the PF
candidates.
After the main GED electron reco is done,
there is a PF "relinking" step which actually connects the PF candidates
and the final gedGsfElectrons.

You may want to discuss details in the EGM meeting
My quick glance at the PFAlgo was suggestive that having a PV with ~3K
tracks may kill  any electrons that could enter the particleFlow collection.

@lgray @Sam-Harper @bachtis



> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/19083#issuecomment-308916152>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblg77BQHjgY8ApO7jlpE6KGBAyijks5sEejCgaJpZM4NugHs>.
>

ttrk commented:
@slava77 Specifically, the electron collection in HI analysis is gedGsfElectronsTmp. I guess this corresponds to the "early" ones.

slava77 commented:
On 6/15/17 7:48 PM, ttrk wrote:
> @slava77 <https://github.com/slava77> Specifically, the electron
> collection in HI analysis is gedGsfElectronsTmp. I guess this
> corresponds to the "early" ones.

yes,
these are before the "PF relinking" step

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/19083#issuecomment-308918555>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbmxrt-253vrVzpY7Nv43g5ros-TRks5sEe0BgaJpZM4NugHs>.
>

cmsbuild commented:
-1

Tested at: 6e4cf2302096c85e11f42f807072f6e27ad910d1

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19083/20627/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
136.731 step1
<pre>DAS Error</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

mandrenguyen commented:
@slava77 I think we decided at some point that there was no good reason not to have the re-linking step in the heavy-ion reco, but I somehow never got around to including it.
I can do that in a subsequent PR.
In any case, we have some work to do to understand how GED electrons behave with the heavy-ion reco in crowded events.

slava77 commented:
@mandrenguyen @abaty 
would you like to move on with this PR, or delay this until the PF electrons issue is understood?
Please clarify.
Thank you.

slava77 commented:
@cmsbuild please test

it should work this time

cmsbuild commented:
The tests are being triggered in jenkins.

mandrenguyen commented:
@slava77 I suggest we move on.

abaty commented:
> If you confirm that you saw chi2/ndof peak around 0.9, then it's expected to have chi2Prob peak at 1

I do see a fairly sharp peak:
![normchi2](https://user-images.githubusercontent.com/5641829/27233814-259e98e8-52bb-11e7-9ff9-4471b27c1111.png)

abaty commented:
@slava77 I agree with Matt, we can move on.

cmsbuild commented:
+1
Tested at: 6e4cf2302096c85e11f42f807072f6e27ad910d1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19083/20636/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19083/20636/summary.html

Comparison Summary:
 - You potentially added 2 lines to the logs
 - Reco comparison results: 18 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1803136
 - DQMHistoTests: Total failures: 29939
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1773031
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/19083	6e4cf23
- the implementation is in line with the description and the follow up review
- jenkins tests pass and comparisons with baseline show differences only in the HI workflow in plots related to the modified vertices
- local tests with 400 events in wf 140.53 as of 8c9ea21 show expected behavior
    - some plots are available in https://github.com/cms-sw/cmssw/pull/19083#issuecomment-308530206 and the behavior is recognized as expected (or to be followed up in the future as with the PF electrons)
    - CPU cost increase is about 4% of the total reco time for events with high multiplicity. The increase is essentially in full from hiOfflinePrimaryVertices taking about 3 s/event  for such events (I simply select the population where this module takes more than 0.1 s/evt: about a half of the events have this and there are no significant outliers, with the mean of about 3 and the maximum of 6.5)
    - hiSelectedVertex branch increases in size from about 500 bytes to about 8.3 kB per event

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

davidlange6 commented:
+1
